### PR TITLE
DrivAerML: lr=4e-4 with full champion EMA+gc stack

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DM champion uses lr=5e-4. On TandemFoil, the optimal LR decreased from 3e-4 to 1.25e-4 to 1e-4 as the config matured with EMA and gc. On AirfRANS, lr=6e-4 won (close to DM). For DM, lr=5e-4 was found during early exploration without EMA or gc — the full champion stack (EMA=0.9995, gc=0.5) may prefer a slightly lower LR since EMA already provides implicit regularization and gc limits gradient magnitude.

**Why lr=4e-4 specifically:**
- gojo #3308 is testing lr=4.8e-4 and 5.2e-4 (narrow sweep around champion)
- lr=4e-4 is a larger step down that probes a qualitatively different optimization regime
- guts #3109 tested lr=4e-4 earlier but likely without EMA+gc (pre-restructure)
- The TF pattern (lower LR + EMA = deeper basins) may transfer to DM

## Instructions

```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 4e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --ema-mode fixed \
  --no-compile-model \
  --save-checkpoint \
  --eval-interval 10 \
  --seed 42 \
  --wandb-group dm-lr-sweep \
  --kill-if-best-val-above 150:8.0 \
  --kill-if-no-improvement 250:0.05
```

**After training**, full-eval on best checkpoint for TEST metric.

## Baseline

- **DM best val:** 3.833% at ep511 (PR #3072, lr=5e-4, W&B: ncl1dh88)
- **DM authoritative TEST:** 4.324%
- **TF LR trend:** 3e-4 → 2e-4 → 1.5e-4 → 1.25e-4 → 1e-4 (each lower LR improved with EMA)
- **DM LR in flight:** gojo #3308 testing lr=4.8e-4/5.2e-4 (narrow sweep)
- **External SOTA:** <3.71% (AB-UPT)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shouko:

## Results

### DrivAerML lr=4e-4 with full champion EMA+gc stack

| Metric | lr=4e-4 (this run) | Champion (lr=5e-4) | Delta |
|--------|-------------------|---------------------|-------|
| **val_primary/surface_rel_l2_pct** | **3.851%** (ep970) | **3.833%** (ep511) | +0.018 pp |
| **test_primary/surface_rel_l2_pct** | **4.428%** | **4.324%** | +0.104 pp |
| External SOTA (AB-UPT) | — | — | <3.71% |

### Details
- **W&B run ID:** zpwubl25
- **W&B group:** dm-lr-sweep
- **Best epoch:** 970 (of 999 max)
- **Peak VRAM:** ~7.4 GB (single GPU)
- **Total training time:** ~9.5 hours (570 min of 600 min timeout)
- **EMA decay:** 0.9995 (fixed mode)
- **Grad clip events:** occasional (EMA smoothed all spikes)

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 4e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --ema-mode fixed \
  --no-compile-model \
  --save-checkpoint \
  --eval-interval 10 \
  --seed 42 \
  --wandb-group dm-lr-sweep \
  --kill-if-best-val-above 150:8.0 \
  --kill-if-no-improvement 250:0.05 \
  --agent shouko \
  --wandb-name "shouko/dm-lr4e4-champion"
```

### What happened

**The hypothesis partially confirmed but did not beat the champion.**

lr=4e-4 with the full EMA+gc stack converged to val=3.851%, just 0.018 pp behind the champion's 3.833%. The convergence trajectory was remarkably slow but steady — the model continued improving through ~33 cosine restart cycles (T_max=30), breaking through multiple apparent plateaus:

- ep430: 4.109% (first stable best)
- ep530: 4.041% (broke through 100-epoch plateau)
- ep600: 3.999% (broke below 4%)
- ep730: 3.907% (broke below 3.95%)
- ep970: 3.851% (final best, 0.26 pp total improvement)

The TandemFoil pattern of "lower LR + EMA = deeper basins" does appear to transfer to DrivAerML — the model found progressively deeper minima through cosine restarts, never truly plateauing even at ep999. However, the lower LR needed **~2x the epochs** (970 vs 511) to approach the champion's val metric, and the val-to-test generalization gap was larger (0.58 pp for lr=4e-4 vs 0.49 pp for lr=5e-4), suggesting the lower LR may overfit slightly more to the validation distribution.

**Key insight:** lr=4e-4 converges more slowly but is still improving at 999 epochs. If given more training budget (2000+ epochs), it might eventually match or beat the champion. The champion at lr=5e-4 appears to have found a better balance between convergence speed and final quality within the given epoch budget.

### Suggested follow-ups

1. **lr=4e-4 with extended training budget** (2000+ epochs) — the model was clearly still improving and may surpass the champion given more time
2. **lr=4.5e-4** — split the difference between champion lr=5e-4 and this run's lr=4e-4 to find the optimal balance of convergence speed and final quality
3. **lr=4e-4 with T_max=50 or T_max=60** — longer cosine periods may help the lower LR make larger steps per cycle
4. **lr=3.5e-4 or lr=3e-4** — test whether even lower LRs with longer training could find deeper basins (though diminishing returns likely)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3318 because ADVISOR: Closing — val=3.851% at lr=4e-4 T_max=30, doesn't beat 3.622% (T_max=36) or 3.700% (metric-aware). Slower convergence without reaching a deeper basin within 999 epochs. The LR optimum at T_max=30 is at 5e-4. Will reassign you to probe LR at T_max=36 MSE-only..
